keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,842 @@
+* Expand the admission surface area (admission.Attributes, AdmissionReview, CEL authorizer implementation available to admission) 
+  since admission verbs don't support field/label selectors
+
+
+## Proposal
+
+List, Watch, and DeleteCollection requests directly have field and label selector options.
+A single-item List or Watch request is still a list as normal (including selectors), but also includes a name.
+
+### Authorization Attributes changes
+
+The authorization attributes have easy access to the query parameter field and label selectors.
+To avoid confusion, field and label selectors will not be included in authorization attributes for verbs where the field
+selector has no semantic meaning.
+In practice this means that (for now), only List, Watch, and DeleteCollection have field and label selectors.
+SubjectAccessReviews submitted to the kube-apiserver with verbs that do not honor the selectors will have the field and
+label selector attributes removed prior to checking authorization.
+Webhook authors: remember that the list of verbs accepting field and label selectors may change over time.
+If the kube-apiserver sends the FieldSelector or LabelSelector to a webhook, the kube-apiserver intends to honor the selector attributes.
+Any authorizer that gets an error from `ParseFieldSelector` or `ParseLabelSelector` may attempt to authorize without
+field or label selectors since that will authorize using a wider permission (field and label selectors can only reduce access).
+
+```go
+type Attributes interface {
+	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
+	// It can return an error if the field selector cannot be parsed.
+    // Remember that field selector formats vary based on the version of the API being used!
+	ParseFieldSelector() ([]Requirement, error)
+  
+	// ParseLabelSelector is lazy, thread-safe, and stores the parsed result and error.
+    // It can return an error if the field selector cannot be parsed.
+    ParseLabelSelector() ([]Requirement, error)
+```
+
+#### Future-proofing your authentication webhook for future verbs
+
+As of 1.31, the only verbs with field and label selectors are List, Watch, and DeleteCollection.
+<<[UNRESOLVED deads2k, liggitt, jpbetz, sttts: this is a future promise for the kube-apiserver behavior if we add it ]>>
+In the future, the kube-apiserver may add field and label selectors to Create, Update, Patch, and Delete.
+<<[/UNRESOLVED]>>
+* For Create, this means that the resource after all mutation is complete (finalObject) must match the field and label selector. 
+* For Update/Patch, this means that the finalNewObject and oldObject must match the field and label selector.
+* For Delete, this means that the oldObject must match the field and label selector.
+ 
+We do not allow field and label selectors for Get, because if a client is specifying a selector, they can add a `.metadata.name`
+field selector and use a List to get equivalent functionality.
```

Comment:
> Are we sure this is good enough?

I can't think of what else Get is beyond a single-item list.

Reply:
Lots of clients are using GET today though. I think there's an assumption here that clients would have to change to inject label selectors anyway, so they might as well switch to single-element LIST, but that might not always be true. I'd also be curious if there's any performance difference between single-element LIST with field selector and GET.

Reply:
> I'd also be curious if there's any performance difference between single-element LIST with field selector and GET.

Yeah this would be interesting to benchmark.

Reply:
Added an explanation of what field and label selectors would do with a GET, but I'm not 100% sold on whether we'd actually add it.